### DISCUSSION TITLE
How to use Lexer Modes in langium service?

### DISCUSSION THREAD

**Discussion by yuyi919** [eclipse-langium/langium#319] [Q&A] [answered]
It can be used to tokenize some markup language

references:
  https://chevrotain.io/docs/features/lexer_modes.html

**Comment by msujew** [comment, marked as answer]
Hi @yuyiRai thank you for your question.

I just wrote up a long-winded answer, until I noticed that our API currently does not allow to construct a multi mode lexer definition...

You can theoretically assign push/pop modes to different token types like this by overriding the [`DefaultTokenBuilder`](https://github.com/langium/langium/blob/d67906aae6a54ec4e9aac8958e85429cdd8c678f/packages/langium/src/parser/token-builder.ts#L18) class. In particular, you can override the `buildKeywordToken` method.

However, there's no way to create a lexer definition like required for Chevrotains lexer because the return type of the `buildTokens` method is an array of tokens. I created a new issue for this #320

**Reply by luiz290788** [reply]
Hi @msujew, I saw that the issue you create is already solved, but the solution is still not clear to me. Would we still need to override the `DefaultTokenBuilder`?

**Reply by msujew** [reply]
Yes, you need to override the `DefaultTokenBuilder` and return a [`IMultiModeLexerDefinition`](https://github.com/Chevrotain/chevrotain/blob/953c35c769f00d83039ac98e5d0ad9cb35b38153/packages/types/api.d.ts#L1743-L1746). We make use of this feature in a few closed source projects.

**Reply by msujew** [reply]
We're currently looking into making it easier to create multi mode lexer, though it's fairly low on our priority list. See also [this discussion](https://github.com/langium/langium/discussions/699) for more information.

**Reply by luiz290788** [reply]
Thanks, it is good to know that this has been validated internally.
